src/main/java/dev/morling/onebrc/CalculateAverage_royvanrijn.java

```diff
@@ -16,64 +16,319 @@
+    private static final long SEPARATOR_PATTERN = compilePattern((byte) ';');
+    private static final long EOL_PATTERN = compilePattern((byte) '\n');
+
+    private BitTwiddledMap processBuffer(ByteBuffer bb) {
+
+        BitTwiddledMap measurements = new BitTwiddledMap();
+
+        final int limit = bb.limit();
+        final byte[] buffer = new byte[64];
+
+        while (bb.position() < limit) {
 
-//        long before = System.currentTimeMillis();
+            final int startPointer = bb.position();
 
-        Map<String, Measurement> resultMap = Files.lines(Path.of(FILE)).parallel()
-                .map(record -> {
-                    // Map to <String,double>
-                    int pivot = record.indexOf(";");
-                    String key = record.substring(0, pivot);
-                    double measured = Double.parseDouble(record.substring(pivot + 1));
-                    return new AbstractMap.SimpleEntry<>(key, measured);
-                })
-                .collect(Collectors.toConcurrentMap(
-                        // Combine/reduce:
-                        AbstractMap.SimpleEntry::getKey,
-                        entry -> new Measurement(entry.getValue()),
-                        Measurement::combineWith));
+            int separatorPointer = findNextSWAR(bb, SEPARATOR_PATTERN, startPointer + 3, limit);
+            int endPointer = findNextSWAR(bb, EOL_PATTERN, separatorPointer + 3, limit);
 
-        System.out.print("{");
-        System.out.print(
-                resultMap.entrySet().stream().sorted(Map.Entry.comparingByKey()).map(Object::toString).collect(Collectors.joining(", ")));
-        System.out.println("}");
+            // Extract the name of the key:
+            final int nameLength = separatorPointer - startPointer;
+            final int entryLength = endPointer - startPointer;
+            final int valueLength = endPointer - separatorPointer - 1;
 
-//        System.out.println("Took: " + (System.currentTimeMillis() - before));
+            // Read the entry in a single get():
+            bb.get(buffer, 0, entryLength);
 
+            // Create the key (no string creation)
+            byte[] key = new byte[nameLength];
+            System.arraycopy(buffer, 0, key, 0, nameLength);
+
+            bb.position(endPointer + 1); // skip the separator
+
+            // Extract the measurement value (10x), skip making a String altogether:
+            final int measured = branchlessParseInt(buffer, nameLength + 1, valueLength);
+
+            measurements.getOrCreate(key).updateWith(measured);
+        }
+
+        return measurements;
     }
+
+    /**
+     * -------- This section contains SWAR code (SIMD Within A Register) which processes a bytebuffer as longs to find values:
+     */
+
+    private int findNextSWAR(ByteBuffer bb, long pattern, int start, int limit) {
+        int i;
+        for (i = start; i <= limit - 8; i += 8) {
+            long word = bb.getLong(i);
```

Comment:
By default ByteBuffers uses Big Indian, which is native for M1/M2...but not on x86, which is the one used for the benchmark.
It translates in:
- wrong position, while found
- a useless reverse bytes 

It would be better to store in a static final, the native byte order and based on it, correctly reverse the bytes yourself (look at what I have done in the Netty code).
Conversely you can always assume little endian, setting the mapped byte buffer order, and leave M1 to have the performance hit

Reply:
Ugh...

Wouldn't it be sufficient to change the numberLeadingZero's to numberTrailingZero's? I believe that's what I Netty's code did. But I'm probably failing to see the difference between getLong/getLongLE?

Reply:
Yep, It Is, but the choice about which order to set the buffer in order to achieve better performance (using native order and save reverse performed by ByteBuffer) force you to add that case to handle both. If you don't care about having a Long.reverse (which is still a single ASM instruction in x86) you can ignore it.
In case you have to add that condition, I suggest to make it static final, which will avoid having the actual condition to happen at runtime.

Reply:
I've decided to just go with LE for now and change the ByteBuffer, that way it should just work.